_posts/2020-01-31-Everything-you-need-to-know-about-ISP.md

```diff
@@ -0,0 +1,166 @@
+3. Availability of methods which client does not use or need confuses the client as well
+4. We will also see further how ISP violation will lead to violation of other SOLID principles as well.
+
+## Code smells for ISP violations and how to fix them
+Whether working solo or in larger teams, it helps to identify problems earlier in code, here are some code smells which could possibly indicate ISP violation:
+
+1. A bulky interface    
+A bulky interface might be a good indicator for ISP violation, although it's not always the case
+
+2. UNSupportedMethod Exception
+As in our burger example, if you encounter too many **Unsupported Operation Exception** it could be a design problem related to ISP. It might be a good time to refactor these classes, for example, we can refactor our BurgerPoint code to have separate interfaces for Burger order and Fries order.
+```java
+/**
+ * IBurgerOrder.java
+ **/
+public interface IBurgerOrder{
+    void orderBurger(int quantity);
+}
+```
+```java
+/**
+ * IFriesOrder.java
+ **/
+public interface IFriesOrder {
+    void orderFries(int quantity);
+}
+```
+You could also use [Adapter pattern](https://en.wikipedia.org/wiki/Adapter_pattern). Adapter patten is used to abstract the unwanted methods like below:
+```java
+/**
+ * IAdapterOrderForBurger.java
+ **/
+interface IAdapterOrderForBurger {
+    void orderBurger(int quantity);
+}
+```
+```java
+/**
+ * IAdapterOrderForFries.java
+ **/
+interface IAdapterOrderForFries {
+    void orderFries(int quantity);
+}
+```
+We have created independent interfaces which will only have the methods needed from the **IOrder.java** interface
+```java
+/**
+ * AdaptedBurgerOrder.java
+ **/
+public class AdaptedBurgerOrder implements IAdapterOrderForBurger {
+    private IOrder burgerOrder;
+    public AdaptedBurgerOrder(IOrder burgerOrder){
+        this.burgerOrder = burgerOrder;
+    }
+
+    @Override
+    public void orderBurger(int quantity) {
+        burgerOrder.orderBurger(quantity);
+    }
+}
+// Similarly for fries order
+```
+As we can see in above code, we are still using the methods provided by **IOrder.java** interface but we are only implementing the methods that the concrete class actually needs.
+This is useful when you are using an external dependency. In the above example  we are using order interface as an external dependency but we have successfully restructured code to avoid side effects. 
+
+3. Too many dependencies to mock for JUnit    
+If your class have too many dependencies while methods only needed few, this might indicate for ISP violation which leads to minimal dependency
+
+4. Too many wrapper classes    
+As wrapper classes add no additional functionality in the code rather is a design choice, having too many of them indicates for refactoring of the codebase. Having too many of wrapper classes will also need maintenance and further increases complexity in codebase
```

Comment:
Yes, but this is rather because of introducing too many wrappers i.e. workaround, indicating refactoring of the codebase, so I have removed this from the code smell section.

Thanks for pointing out